Fix consumption not updated on push

andig opened the pull request:
Fix https://github.com/volkszaehler/volkszaehler.org/issues/381, part one- no consumption update

frankrichter commented:
Pragmatischer Vorschlag zur Korrektur von consumption um die nach links weggeschobenen Tupel: 
```javascript
					// update consumption
					consumption /= 3.6e6;
					if (this.data.consumption !== undefined) {
						this.data.consumption = (this.data.consumption || 0) + consumption;

						// calculate new left plot border and remove outdated tuples and consumption
						var left = deltaTuples[deltaTuples.length-1][0] - vz.options.plot.xaxis.max + vz.options.plot.xaxis.min;
						while (this.data.tuples[0][0] < left) {
							var first = this.data.tuples[0];
							this.data.consumption -= first[1] * (first[0] - this.data.from) / 3.6e6;
							this.data.from = first[0];
							this.data.tuples.shift();
						}
					}

					if (this.initialconsumption !== undefined) {
						this.totalconsumption = (this.totalconsumption || 0) + consumption;
					}

```
Perfekt ist es nicht, weil kleine Sprünge entstehen (es kommen ständig kleine Häppchen dazu, dafür wird ab und zu ein größerer Brocken abgezogen). Aber der angezeigte Verbrauch ist nur noch maximal ein MW-Tupel daneben und wird nicht mehr nur akkumuliert.

andig commented:
@frankrichter wie immer sehr guter Vorschlag. PR aktualisiert.

andig commented:
ping @frankrichter test ok?

frankrichter commented:
Sorry, mir war nicht klar dass du auf Test wartest. Hab jetzt deine Anpassungen noch in meine Installation übernommen, läuft wie erwartet.

andig commented:
>  mir war nicht klar dass du auf Test wartest.

Immer besser da ich keinen Pushserver laufen habe- vielen Dank! 

**OT:** Ich schaue allerdings gerade wie der PS sich mit der High Perfromance Middleware (ppm) verbinden lässt so dass _alle_ Middleware Updates auch als Pushnachrichten verteilt werden. Damit würden dann auch andere Erzeuger als vzlogger prinzipiell Echtzeitupdates generieren können.

frankrichter commented:
Hört sich interessant an. Dann muss ich mich wohl auch mal mit ppm auseinandersetzen...
Zu push hätte ich auch Frontend-seitig noch einige Ideen. Hast du da Bock drauf oder eher nicht so, wenn du keine Testmöglichkeit und keine eigene Verwendung hast?

frankrichter commented:
consumption update bei aktivem push erfolgt jetzt leider immer, also auch für `!vz.wui.tmaxnow`. Das sollten wir noch irgendwie verhindern.

andig commented:
> consumption update bei aktivem push erfolgt jetzt leider immer,

Mhm. Wenn ich den PR so anschaue müsste das eigentlich schon immer der Fall gewesen sein?

> also auch für !vz.wui.tmaxnow. Das sollten wir noch irgendwie verhindern.

Yep. Könntest Du mal probieren ob sich das nicht in dieser Zeile hier mit abfrühstücken ließe: https://github.com/volkszaehler/volkszaehler.org/pull/610/files#diff-b3ca937ef7d24c22606ae1ce157d935eL194

frankrichter commented:
> Mhm. Wenn ich den PR so anschaue müsste das eigentlich schon immer der Fall gewesen sein?

Es war eigentlich immer so angelegt, ja. Aber weil es wegen falscher JSON-Struktur (1. commit) nicht funktionierte, ist es nicht aufgefallen.
Mit meinem 2. Vorschlag (remove outdated) hab ich das Problem dann noch verschlimmert, denn wenn man einen Zeitraum in der Vergangenheit lädt und das erste Push-Update eintrifft bevor der Graph gezeichnet ist räumt die while-loop alle Tupel ab und hinterher fehlt der Graph...
Hat mich gestern einiges Kopfzerbrechen gekostet bis ich gemerkt habe dass es an push liegt. 

> Könntest Du mal probieren ob sich das nicht in dieser Zeile hier mit abfrühstücken ließe: https://github.com/volkszaehler/volkszaehler.org/pull/610/files#diff-b3ca937ef7d24c22606ae1ce157d935eL194

Test folgt.

frankrichter commented:
Ja, das klappt so. In Zeile 194 steht bei mir jetzt:
`if (!push.data || push.data.uuid !== this.uuid || !vz.wui.tmaxnow) {`

Nebenwirkung: In der "aktuell"-Spalte stand bisher immer der letzte Push-Wert, jetzt ist es logischerweise wieder das letze Tupel des MW-Requests wie bei non-push-Kanälen. Ist aber so auch ok, ob der Live-Wert wirklich in eine Tabelle mit historischen Daten reingehört war zumindest diskussionswürdig.

andig commented:
Schau mal ob Du für die anderen Fehler mit diesem Diff glücklich wirst (an push nix geändert):

*** /var/folders/w5/_c0nb6n90fn96tzw04dtc6240000gn/T/JwMeXa_wui.js  2017-07-08 11:35:13 +0000
--- /Users/andig/Documents/htdocs/vz/htdocs/js/wui.js 2017-07-08 11:34:19 +0000
***************
*** 57,69 ****
      // bind display mode change
    $('#display select').change(function(event) {
      var val = $(this).val().split('-');
!     vz.wui.changeDisplayMode(val.pop());
      if (val.length) {
!       vz.wui.handleControls('zoom-' + val.pop());
      }
      else {
!       vz.wui.zoom(vz.options.plot.xaxis.min, vz.options.plot.xaxis.max);
      }
    });
  
--- 57,72 ----
      // bind display mode change
    $('#display select').change(function(event) {
      var val = $(this).val().split('-');
!     var mode = val.pop();
!     var keepPeriodStartFixed = vz.options.mode !== mode;
!     vz.wui.changeDisplayMode(mode);
      if (val.length) {
!       vz.wui.handleControls('zoom-' + val.pop(), keepPeriodStartFixed);
      }
      else {
!       var to = Math.min(new Date().getTime(), vz.options.plot.xaxis.max);
!       vz.wui.zoom(vz.options.plot.xaxis.min, to);
      }
    });
  
***************
*** 570,577 ****
   *
   * @param ev either click event or literal button event value
   */
! vz.wui.handleControls = function(action) {
    var delta = vz.options.plot.xaxis.max - vz.options.plot.xaxis.min,
        middle = vz.options.plot.xaxis.min + delta/2,
        startOfPeriodLocale;
--- 573,580 ----
   *
   * @param ev either click event or literal button event value
   */
! vz.wui.handleControls = function(action, keepPeriodStartFixed) {
    var delta = vz.options.plot.xaxis.max - vz.options.plot.xaxis.min,
        middle = vz.options.plot.xaxis.min + delta/2,
        startOfPeriodLocale;
***************
*** 656,663 ****
  
        if (vz.wui.tmaxnow) {
          /* jshint laxbreak: true */
!         min = period === vz.wui.period
            ? moment().subtract(1, period).valueOf()
            : moment().startOf(periodLocale).valueOf();
          max = moment().valueOf();
--- 659,666 ----
  
        if (vz.wui.tmaxnow) {
          /* jshint laxbreak: true */
!         min = period === vz.wui.period &! keepPeriodStartFixed
            ? moment().subtract(1, period).valueOf()
            : moment().startOf(periodLocale).valueOf();
          max = moment().valueOf();


Viele Grüße, Andreas

> On 7. Jul 2017, at 18:31, frankrichter <notifications@github.com> wrote:
> 
> Ja, das klappt so. In Zeile 194 steht bei mir jetzt:
> if (!push.data || push.data.uuid !== this.uuid || !vz.wui.tmaxnow) {
> 
> Nebenwirkung: In der "aktuell"-Spalte stand bisher immer der letzte Push-Wert, jetzt ist es logischerweise wieder das letze Tupel des MW-Requests wie bei non-push-Kanälen. Ist aber so auch ok, ob der Live-Wert wirklich in eine Tabelle mit historischen Daten reingehört war zumindest diskussionswürdig.
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/volkszaehler/volkszaehler.org/pull/610#issuecomment-313730385>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AALR72nZjYgQuUfwfzS9xUZUBkTw2SbOks5sLl1bgaJpZM4NDbBD>.
>

andig commented:
Fixed via 7e6b0aa0c6f4ea1a892cd76822088a09abdf20da
